dnscrypt.c

```diff
@@ -3,19 +3,29 @@
 const KeyPair *
 find_keypair(const struct context *c,
              const unsigned char magic_query[DNSCRYPT_MAGIC_HEADER_LEN],
-             const size_t dns_query_len)
+             const size_t dns_query_len,
+             bool *use_xchacha20)
 {
+    unsigned char magic_query_ref[DNSCRYPT_MAGIC_HEADER_LEN];
     const KeyPair *keypairs = c->keypairs;
     size_t i;
 
     if (dns_query_len <= DNSCRYPT_QUERY_HEADER_SIZE) {
         return NULL;
     }
+    *use_xchacha20 = 0;
     for (i = 0U; i < c->keypairs_count; i++) {
-        if (memcmp(keypairs[i].crypt_publickey, magic_query,
-                   DNSCRYPT_MAGIC_HEADER_LEN) == 0) {
+        memcpy(magic_query_ref, keypairs[i].crypt_publickey, sizeof magic_query_ref);
+        if (memcmp(magic_query_ref, magic_query, DNSCRYPT_MAGIC_HEADER_LEN) == 0) {
             return &keypairs[i];
         }
+#ifdef HAVE_CRYPTO_BOX_CURVE25519XCHACHA20POLY1305_OPEN_EASY
+        sodium_increment(magic_query_ref, sizeof magic_query_ref);
+        if (memcmp(magic_query_ref, magic_query, DNSCRYPT_MAGIC_HEADER_LEN) == 0) {
+            *use_xchacha20 = 1;
```

Comment:
so if I understand correctly, `if client_magic == pk[:8] + 1`, then it is assumed that we use xchacha20.
If I am correct, this seems a bit brittle though.

This seems very arbitrary and I failed to find anything in the [spec](https://github.com/jedisct1/dnscrypt-proxy/blob/master/DNSCRYPT-V2-PROTOCOL.txt).
We should be able to match a key with a cert and store as part of the keypair if whether or not xchacha20 should be used right? Basically storing the ES version as part of the keypair. This should also allow the rest of the program to know if salsa20 or chacha20 should be used without the need to pass the argument all along.

@jedisct1 would that work?